src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
 }
 
 template <typename DataType, typename DistType>
-idType HNSWIndex<DataType, DistType>::mutuallyConnectNewElement(
-    idType cur_c, candidatesMaxHeap<DistType> &top_candidates, size_t level) {
-    size_t Mcurmax = level ? maxM_ : maxM0_;
-    getNeighborsByHeuristic2(top_candidates, M_);
-    if (top_candidates.size() > M_)
-        throw std::runtime_error(
-            "Should be not be more than M_ candidates returned by the heuristic");
+void HNSWIndex<DataType, DistType>::revisitNeighborConnections(
+    size_t level, idType new_node_id, idType selected_neighbor, idType *new_node_neighbors_list,
+    idType *neighbor_neighbors_list, std::unique_lock<std::mutex> &node_lock,
+    std::unique_lock<std::mutex> &neighbor_lock) {
 
-    vecsim_stl::vector<idType> selectedNeighbors(this->allocator);
-    selectedNeighbors.reserve(M_);
-    while (top_candidates.size() > 0) {
-        selectedNeighbors.push_back(top_candidates.top().second);
-        top_candidates.pop();
+    // Collect the existing neighbors and the new node as the neighbor's neighbors candidates.
+    candidatesMaxHeap<DistType> candidates(this->allocator);
+    DistType dist_cur_node_neighbor = this->dist_func(
+        getDataByInternalId(new_node_id), getDataByInternalId(selected_neighbor), this->dim);
+    candidates.emplace(dist_cur_node_neighbor, new_node_id);
```

Comment:
You should have this distance already calculated. consider using it and save recalculation